kratos/containers/container_expression/expressions/expression.cpp

```diff
@@ -20,6 +20,72 @@
 
 namespace Kratos {
 
+Expression::ExpressionIterator::ExpressionIterator(Expression::Pointer pExpression)
+    : mpExpression(pExpression),
+      mEntityIndex(0),
+      mEntityDataBeginIndex(0),
+      mComponentIndex(0),
+      mFlattenedShapeSize(pExpression->GetFlattenedShapeSize())
+{
+}
+
+Expression::ExpressionIterator::ExpressionIterator(const ExpressionIterator& rOther)
+    : mpExpression(rOther.mpExpression),
+      mEntityIndex(rOther.mEntityIndex),
+      mEntityDataBeginIndex(rOther.mEntityDataBeginIndex),
+      mComponentIndex(rOther.mComponentIndex),
+      mFlattenedShapeSize(rOther.mFlattenedShapeSize)
+{
+}
+
+Expression::Pointer Expression::ExpressionIterator::GetExpression() const
+{
+    return mpExpression;
+}
+
+double Expression::ExpressionIterator::operator*() const
+{
+    return mpExpression->Evaluate(mEntityIndex, mEntityDataBeginIndex, mComponentIndex);
+}
+
+bool Expression::ExpressionIterator::operator==(const ExpressionIterator& rOther) const
+{
+    return (&*mpExpression == &*rOther.mpExpression) && (mEntityIndex == rOther.mEntityIndex) && (mComponentIndex == rOther.mComponentIndex);
+}
+
+bool Expression::ExpressionIterator::operator!=(const ExpressionIterator& rOther) const
+{
+    return !this->operator==(rOther);
+}
+
+Expression::ExpressionIterator& Expression::ExpressionIterator::operator=(const ExpressionIterator& rOther)
+{
+    mpExpression = rOther.mpExpression;
+    mEntityIndex = rOther.mEntityIndex;
+    mEntityDataBeginIndex = rOther.mEntityDataBeginIndex;
+    mComponentIndex = rOther.mComponentIndex;
+    mFlattenedShapeSize = rOther.mFlattenedShapeSize;
+    return *this;
+}
+
+Expression::ExpressionIterator& Expression::ExpressionIterator::operator++()
+{
+    ++mComponentIndex;
+    if (mComponentIndex == mFlattenedShapeSize) {
+        mComponentIndex = 0;
+        ++mEntityIndex;
+        mEntityDataBeginIndex = mEntityIndex * mFlattenedShapeSize;
+    }
+    return *this;
```

Comment:
I'm not blocking but I just have to note here for the record that I still believe an implementation in terms of an absolute index would be cleaner.